### DISCUSSION TITLE
Custom field resolver

### DISCUSSION THREAD

**Discussion by ronjunevaldoz** [ExpediaGroup/graphql-kotlin#1417] [General]
Is there any existing example for custom field resolver?

**Comment by dariuszkuc** [comment]
Hello 👋 
Can you elaborate on your question? i.e. we generate schema from the source code using reflections so your functions == resolver for that field (properties are just read from object).

**Reply by ronjunevaldoz** [reply]
Hi 
Let's say we have a user data class that only consists of `firstName` and `lastName` I want to dynamically add an additional field to the graphQL response, like fullName as a custom field that is not added to the User.

```kotlin
data class User(val firstName: String, lastName: String)

class UserQuery : Query {
    suspend fun getUserById(userId: String) {
          return User(
                firstName = "Juan",
                lastName = "Dela Cruz"
          )
    }
}
```

## Expected feature
Example feature, which has custom field name and a field type, now everytime I use User class as response, fullName field will automatically added to it.
```kotlin
registerResolver<User>("fullName", String::class) { user->
     return "${user.firstName} ${user.lastName}"
}
```

**Reply by samuelAndalon** [reply]
You can write as many indirection layers as you want, probably you can do something like this

```kotlin

data class User(val firstName: String, lastName: String)

class UserWrapper(val user: User) {
  fun getName(): String = user.firstName
  fun lastName(): String = user.lastName
  fun getFullName(): String = "${user.firstName} ${user.lastName}"
}

class UserQuery : Query {
    suspend fun getUserById(userId: String) {
          return UserWrapper(
            User(
                firstName = "Juan",
                lastName = "Dela Cruz"
            )
          )
    }
}
```

this kind of matches what a DataFetcher is intended to do

```kotlin
TypeRuntimeWiring.newTypeWiring("User")
  .dataFetcher("fullName") { environment ->
     val user = environment.getSource<User>()
     "${user.firstName} ${user.lastName}"
}
```

The key point is that your domain object does not necessarily have to map your schema as you can write as many indirection layers that you need that can help to describe your domain objects.
